ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Types.hs

```diff
@@ -0,0 +1,204 @@
+  FieldType Width = Int
+  FieldType Height = Int
+  FieldType Padding = Int
+  FieldType Margin = Int
+  FieldType Color = Text
+  FieldType BackgroundColor = Text
+  FieldType BorderColor = Text
+  FieldType BorderWidth = Int
+  FieldType BorderRadius = Int
+  FieldType BorderStyle = BorderStyleValue
+  FieldType FontStyle = FontStyleValue
+  FieldType FontWeight = FontWeightValue
+  FieldType FontSize = Int
+  FieldType FontFamily = Text
+
+  -- StringClass
+  FieldType StringValue = Text
+  FieldType Placeholder = Text
+
+  -- Other widgets
+  FieldType Description = Text
+  FieldType ClickHandler = Text -> IO ()
+  FieldType Disabled = Bool
+
+type family Append (xs :: [k]) (ys :: [k]) :: [k] where
+  Append '[] xs = xs
+  Append xs '[] = xs
+  Append (x ': xs) ys = x ': Append xs ys
+
+data WidgetClasses = WidgetClass
+                   | DOMWidgetClass
+                   | StringClass
+
+type family ClassFields (c :: WidgetClasses) :: [Field] where
+  ClassFields WidgetClass =
+    '[ModelModule, ModelName, ViewModule, ViewName, MsgThrottle, Version, OnDisplayed]
+  ClassFields DOMWidgetClass = Append (ClassFields WidgetClass)
+    '[ Visible, CSS, DOMClasses, Width, Height, Padding, Margin, Color
+     , BackgroundColor, BorderColor, BorderWidth, BorderStyle, FontStyle
+     , FontWeight, FontSize, FontFamily
+     ]
+  ClassFields StringClass = Append (ClassFields WidgetClass)
+    '[StringValue, Disabled, Description, Placeholder]
+
+data WidgetType = ButtonType
+                | ImageType
+                | OutputType
+                | HTMLType
+                | LatexType
+                | TextType
+                | TextAreaType
+
+data Widget (w :: WidgetType) = Widget { attrs :: IORef (M.Map Field Dynamic) }
+
+type family WidgetFields (w :: WidgetType) :: [Field] where
+  WidgetFields ButtonType = Append (ClassFields DOMWidgetClass)
+    '[Description, Tooltip, Disabled, Icon, ButtonStyle, ClickHandler]
+  WidgetFields ImageType = Append (ClassFields DOMWidgetClass) '[ImageFormat, B64Value]
+  WidgetFields OutputType = ClassFields DOMWidgetClass
+  WidgetFields HTMLType = ClassFields StringClass
+  WidgetFields LatexType = ClassFields StringClass
+  WidgetFields TextType = ClassFields StringClass
+  WidgetFields TextAreaType = ClassFields StringClass
+
+type family IsElem (xs :: [a]) (x :: a) :: Bool where
+  IsElem '[] y = False
+  IsElem (y ': ys) y = True
+  IsElem (x ': ys) y = IsElem ys y
+
+setField :: (Typeable (FieldType f), SingI f, IsElem (WidgetFields w) f ~ True) => Widget w -> Proxy f -> FieldType f -> IO ()
+setField widget proxy val = do
+  let field = reflect proxy
+  modifyIORef (attrs widget) (M.insert field $ toDyn val)
+
+getField :: (Typeable (FieldType f), SingI f, IsElem (WidgetFields w) f ~ True) => Widget w -> Proxy f -> IO (FieldType f)
+getField widget proxy = do
+  let field = reflect proxy
+  valMap <- readIORef $ attrs widget
+  case M.lookup field valMap of
+    Nothing -> return undefined
```

Comment:
It *should* not happen, if the widgets were constructed properly and the properties were not modified. It will hopefully be well hidden to prevent such editing.